colossalai/zero/gemini/gemini_optimizer.py

```diff
@@ -450,8 +465,14 @@ def collect_states(self, param_id: int, only_rank_0: bool = True) -> dict:
                             states["step"], dtype=torch.float32, requires_grad=False
                         ).cpu()
                     else:
-                        state_tensor = states[state_name].detach().clone().to(torch.float32).cpu()
-                        collected_states[state_name] = torch.reshape(state_tensor, param.shape)
+                        state_tensor = states[state_name].detach().clone().to(torch.float32).cuda()
+                        state_tensor = state_tensor.view(current_shape)
+                        if partition_dim is not None:
+                            gather_tensor = [torch.zeros_like(state_tensor) for _ in range(self.tp_size)]
```

Comment:
Do you consider fused MLP layer?

Reply:
Yes,  the gathering is based on distributed tensors and can also handle fused MLP layers.

Reply:
Can you reuse dtensor api?